src/Surfnet/StepupGateway/ApiBundle/Service/U2fVerificationService.php

```diff
@@ -0,0 +1,256 @@
+        } elseif ($result->didDeviceReportAnyError()) {
+            $apiResult->status = U2fRegistrationVerificationResult::STATUS_DEVICE_ERROR;
+            $this->logger->error('U2F device reported an error');
+        } elseif ($result->didResponseChallengeNotMatchRequestChallenge()) {
+            $apiResult->status = U2fRegistrationVerificationResult::STATUS_UNMATCHED_REGISTRATION_CHALLENGE;
+            $this->logger->error('Response challenge did not match request challenge');
+        } elseif ($result->wasResponseNotSignedByDevice()) {
+            $apiResult->status = U2fRegistrationVerificationResult::STATUS_RESPONSE_NOT_SIGNED_BY_DEVICE;
+            $this->logger->error('Response was not signed by device');
+        } elseif ($result->canDeviceNotBeTrusted()) {
+            $apiResult->status = U2fRegistrationVerificationResult::STATUS_UNTRUSTED_DEVICE;
+            $this->logger->error('The device\'s certificate can not be trusted');
+        } elseif ($result->didPublicKeyDecodingFail()) {
+            $apiResult->status = U2fRegistrationVerificationResult::STATUS_PUBLIC_KEY_DECODING_FAILED;
+            $this->logger->error('Decoding of the public key failed');
+        } elseif ($result->didntAppIdsMatch()) {
+            $apiResult->status = U2fRegistrationVerificationResult::STATUS_APP_ID_MISMATCH;
+            $this->logger->critical('The AppID of a U2F message didn\'t match the server\'s');
+        } else {
+            throw new LogicException('Unknown registration verification result status');
+        }
+
+        return $apiResult;
+    }
+
+    /**
+     * Request signing of a sign request. Does not support U2F's sign counter system.
+     *
+     * @param U2fSignRequest  $request The sign request that you requested earlier and was used to query the U2F device.
+     * @param U2fSignResponse $response The response that the U2F device gave in response to the sign request.
+     * @param Requester    $requester
+     * @return U2fAuthenticationVerificationResult
+     */
+    public function verifyAuthentication(U2fSignRequest $request, U2fSignResponse $response, Requester $requester)
+    {
+        $this->logger->notice('Received request to verify a U2F device registration');
+
+        try {
+            $result = $this->verificationService->verifyAuthentication(
+                $this->adaptSignRequest($request),
+                $this->adaptSignResponse($response)
+            );
+        } catch (Exception $e) {
+            $errorMessage = sprintf(
+                'An exception was thrown while verifying the U2F device authentication (%s: %s)',
+                get_class($e),
+                $e->getMessage()
+            );
+            $this->logger->critical($errorMessage, ['exception' => $e]);
+
+            throw new RuntimeException($errorMessage, 0, $e);
+        }
+
+        $apiResult = new U2fAuthenticationVerificationResult();
```

Comment:
seems this code (and it's siblings in other places :wink:) can be made quite a lot simpler with something like:

```php
<?php

// includes mapping from $result status to internal status, unknown result leads to LogicException
$apiResult = U2fAuthenticationVerificationResult::fromVerificationResult($result);

if (!$result->wasSuccussful()) {
    $this->logger->error(sprintf('U2F authentication verification failed, reason: "%s"', $apiResult->getStatus()));
} else {
    $this->logger->notice('U2F authentication verification successful');
}

return $apiResult
```